wip-http-post.md

```diff
@@ -0,0 +1,138 @@
+  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
+  Status: Draft
+  Type: Standards Track
+  Created: 2021-12-02
+  License: BSD-2-Clause
+</pre>
+
+
+## Abstract
+
+RADON supports HTTP retrieval sources, but only using the HTTP verb GET. Some APIs require the use of the HTTP verb POST. This proposal specifies support for HTTP POST requests.
+
+## Motivation and rationale
+
+While HTTP-GET requests can cover the majority of use cases, there are some APIs that require the use of POST requests. The most common example is the case of GraphQL APIs, which use a POST request with a JSON body to specify what kind of data will be retrieved.
+
+## Specification
+
+At the protocol level, these are the new fields in the protobuf schema:
+
+```diff
+ enum RADType {
+     Unknown = 0;
+     HttpGet = 1;
+     Rng = 2;
++    HttpPost = 3;
+ }
+ message RADRetrieve {
+     RADType kind = 1;
+     string url = 2;
+     // TODO: RADScript should maybe be a type?
+     bytes script = 3;
++    // Body of HTTP-POST request
++    bytes body = 4;
++    // Extra headers for HTTP-GET and HTTP-POST requests
++    repeated StringPair headers = 5;
+ }
+```
+
+#### HTTP-POST
+
+This retrieval type is similar to HTTP-GET, except:
+
+* Using POST instead of GET.
+* Allowing an additional `body` parameter, which is an array of bytes that will be sent to the HTTP server.
+
+#### New `body` field
+
+When the `RADType` is HTTP-GET or RNG, the `body` field must be empty.
+
+When the `RADType` is HTTP-POST, the `body` field may contain an arbitrary sequence of bytes.
+
+The content of this field will be sent to the HTTP server as part of the POST request.
+
+#### New `headers` field
+
+When the `RADType` is RNG, the `headers` field must be empty.
+
+When the `RADType` is HTTP-GET or HTTP-POST, the `body` field may contain a list of `(key, value)` pairs, using the `StringPair` helper message:
+
+```
+message StringPair {
+    string left = 1;
+    string right = 2;
+}
+```
+
+These pairs will be used as additional headers for the HTTP request. At the protocol level they can be any string, but the RAD engine will validate them according to the HTTP specification, and resolve the retrieval source to a `RadonError` in case of error.
+
+Since HTTP-GET requests follow a very similar structure to HTTP-POST requests, the addition of the `headers` field will also be implemented in the HTTP-GET handler. The semantics are exactly the same as in HTTP-POST requests.
+
+#### Updated transaction weights
+
+The new fields of the `RADRetrieve` message must be taken into account when calculating the data request transaction weight.
+
+Old weight:
+
+```
+RADRetrieve weight = (script bytes) + (url bytes) + (1)
+```
+
+New weight:
+
+```
+RADRetrieve weight = (script bytes) + (url bytes) + (1) + (body bytes) + (headers_weight)
+headers_weight = for each header: (key bytes) + (value bytes) + overhead
+overhead = 2 (?) (because of StringPair)
```

Comment:
Why the `overhead` is needed?

Reply:
Without the overhead an attacker could submit a list of headers such as:

```
[("", ""), ("", ""), ("", "")]
```

for free, because the length of each header key and value is 0.

Reply:
Interesting. Sounds reasonable to me.

Reply:
```suggestion
overhead = 2
```

I'm OK with this as long as we add an explanation below.

Reply:
The goal is to match the serialized size as close as possible, I haven't calculated the exact overhead yet because I wasn't sure if the final implementation would use the `StringPair` type.

Reply:
So actually the overhead is `6`, because protocol buffers implements `repeated` by simply repeating the field. So for every header there are 2 bytes to indicate this is a header, 2 bytes to indicate "this is a string with length n" before the header name, 2 bytes to indicate "this is a string with length n" before the header value.